src/app/outbound/resolve.rs

```diff
@@ -0,0 +1,110 @@
+use super::endpoint;
+use crate::api_resolve::Metadata;
+use crate::app::dst::DstAddr;
+use crate::dns::Suffix;
+use linkerd2_error::{Error, Recover};
+use linkerd2_exp_backoff::{ExponentialBackoff, ExponentialBackoffStream};
+use linkerd2_proxy_core::{resolve, Resolve};
+use linkerd2_proxy_resolve::{map_endpoint, recover};
+use linkerd2_request_filter as request_filter;
+use std::sync::Arc;
+use tower_grpc as grpc;
+
+pub fn resolve<R>(
+    suffixes: Vec<Suffix>,
+    backoff: ExponentialBackoff,
+    resolve: R,
+) -> map_endpoint::Resolve<
+    endpoint::FromMetadata,
+    request_filter::Service<
+        PermitNamesInSuffixes,
+        resolve::Service<recover::Resolve<BackoffUnlessInvalidArgument, R>>,
+    >,
+>
+where
+    R: Resolve<DstAddr, Endpoint = Metadata> + Clone,
+{
+    map_endpoint::Resolve::new(
+        endpoint::FromMetadata,
+        request_filter::Service::new(
+            suffixes.into(),
+            recover::Resolve::new(backoff.into(), resolve).into_service(),
+        ),
+    )
+}
+
+#[derive(Clone, Debug)]
+pub struct PermitNamesInSuffixes {
+    permitted: Arc<Vec<Suffix>>,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct BackoffUnlessInvalidArgument(ExponentialBackoff);
+
+#[derive(Debug)]
+pub struct Unresolvable(());
+
+// === impl PermitNamesInSuffixes ===
+
+impl From<Vec<Suffix>> for PermitNamesInSuffixes {
+    fn from(permitted: Vec<Suffix>) -> Self {
+        Self {
+            permitted: Arc::new(permitted),
+        }
+    }
+}
+
+impl request_filter::RequestFilter<DstAddr> for PermitNamesInSuffixes {
+    type Error = Unresolvable;
+
+    fn filter(&self, dst: DstAddr) -> Result<DstAddr, Self::Error> {
+        if let Some(name) = dst.dst_concrete().name_addr() {
+            if self
+                .permitted
+                .iter()
+                .any(|suffix| suffix.contains(name.name()))
+            {
+                tracing::debug!("suffix matches");
```

Comment:
Nit: should this include what the matched suffix was? or what the destination name was?